engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,167 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of pausing/stopping blocks execution.
+type StopControl struct {
+	sync.RWMutex
+	height             uint64
+	crash              bool
+	commenced          bool // if stopping process has started. We disallow any changes then.
+	set                bool // whether stop at height has been set at all, it its envisioned most of the time it won't
+	paused             bool // if execution is paused
+	stopAfterExecuting flow.Identifier
```

Comment:
I think we can get rid of the `paused` state variable. Here is why.

I think the minimum state set is:

```golang
// on startup, this is initialized as highest executed height
// after startup, this is updated by a Processable block with higher height 
// stop height can not be set below or equal to this height, otherwise we can't guarantee 
// no block above stopHeight has been executed. 
highestProcessableHeight uint64

// this is true means (height, crash) are set
stopHeightSet bool 

// the height above (or equal to) which won't be executed
height uint64 

// whether to crash when (height - 1) is finalized and executed. 
crash bool 

// indicate if any block above lastHeight has been seen. lastHeight is (stopHeight - 1).
// this is set to true when BlockProcessable is called with a block whose height is above lastHeight
// if this happens, stop height can not be changed, because we might have stopped some block
// from executing.
hasBlockAboveLastHeight bool 

// whether block at (height - 1) is finalized. lastHeight is (stopHeight - 1).
lastHeightFinalized bool 

// the ID for the finalized block at (height - 1), only updated when `lastHeightFinalized` is true.
lastHeightID flow.Identifier 

// indicate whether the last height (the block with lastHeightID) has been executed,
// note must be the finalized block at last height, not a different block at the same height.
lastHeightExecuted bool
```

Now with the above state variable, the most important value is the return value of `BlockProcessable`, which decides whether to execute that block. 

Is suggest this function to return true for "processable", instead of "skip". As it's easy to read:
```
// Notify the stop control about a processable block and return whether the block is processable.
func (s *StopControl) NotifyBlockProcessable(h *flow.Header) bool

// in ingestion engine
func (e *IngestionEngine) BlockProcessable(h *flow.Header) {
  processable := s.NotifyBlockProcessable(h) 
  if !processable {
    // queue the block to process 
  }
}
```

And NotifyBlockProcessable can be implemented as:
```
func (s *StopControl) NotifyBlockProcessable(h *flow.Header) bool {
  s.Lock()
  defer s.Unlock()

  // if paused already, then we won't execute any block
  if s.paused() {
    return false
  }
  
  // always update the highest height
  if h.Height > s.highestProcessableHeight {
    s.highestProcessableHeight = h.Height
  }

  // if stop height not set, any block are executable
  if !set {
    return true
  }

  lastHeight := s.height - 1
  if h.Height > lastHeight {
    s.hasBlockAboveLastHeight = true 
  }
}

func (s *StopControl) paused() bool {
  return s.lastHeightFinalized && s.lastHeightExecuted
}
```

Reply:
Now you might wonder what if user want to pause the execution to begin with?

That's actually what I wanted to distinguish. 

The `pause` state means two things:
1. user wants to pause the execution to begin with
2. user wants to not crash after reaching stop height.

For 2, we can use the StopControl module without a pause method, rather than a state
For 1, we can use a AlwaysSkip module which implements the same interface as StopControl, but the NotifyBlockProcessable will always return false (not processable). 

This way, we get rid of the `pause` state from `StopControl` module.

Reply:
I like the idea of tracking latest height, and preventing setting new one if thats below or close to it.
However, I think path with interfaces and separate modules for rather simple distinction is a bit of an overkill.

Reply:
The pause flag is a like a main control above all other controls. 

If you don't like using interface, consider wrapping StopControl with an addition pause flag:
```
type StopController struct {
  pause bool
  stopControl *stopControl
}
```